### DISCUSSION TITLE
Create executable with Open3D using PyInstaller

### DISCUSSION THREAD

**Discussion by hrimov** [pyinstaller/pyinstaller#8256] [Help] [answered]
I have a desktop application in PyQt6 that uses Open3D for point cloud visualisation. I tried to bundle it to the standalone executable using PyInstaller, but it crashes when I'm trying to run it with the DLL load failed error(traceback below). 
```
Traceback (most recent call last):
   File "src\__main__.py", line 5, in <module>
   File "PyInstaller\loader\pyimod02_importers.py", line 385, in exec_module
   File "src\__init__.py", line 1, in <module>
   File "PyInstaller\loader\pyimod02_importers.py", line 385, in exec_module
   File "src\__main__.py", line 5, in <module>
   File "PyInstaller\loader\pyimod02_importers.py", line 385, in exec_module
   File "src\ui\__init__.py", line 1, in <module>
   File "PyInstaller\loader\pyimod02_importers.py", line 385, in exec_module
   File "src\ui\main.py", line 13, in <module>
   File "PyInstaller\loader\pyimod02_importers.py", line 385, in exec_module
   File "src\ui\views\presenter.py", line 25, in <module>
   File "PyInstaller\loader\pyimod02_importers.py", line 385, in exec_module
   File "src\ui\views\components\video\open3d_.py", line 1, in <module>
   File "PyInstaller\loader\pyimod02_importers.py", line 385, in exec_module
   File "open3d\__init__.py", line 97, in <module>
ImportError: DLL load failed while importing pybind: A dynamic link library (DLL) initialization routine failed.
[3924] Failed to execute script '__main__' due to unhandled exception!
```
It happens only when I try to run it on the Windows device that do not have C++ Build tools. So, the question is how can I create standalone executable without the additional step with the build tools installation?

I have the following `.spec` file:
```
# -*- mode: python ; coding: utf-8 -*-


block_cipher = None


a = Analysis(
    ['src\\__main__.py'],
    pathex=[],
    binaries=[],
    datas=[('config\\\\*', 'config'), ('src\\\\ui\\\\design\\\\styles', 'src\\\\ui\\\\design\\\\styles'), ('src\\\\ui\\\\design\\\\fonts', 'src\\\\ui\\\\design\\\\fonts'), ('src\\\\ui\\\\design\\\\icons', 'src\\\\ui\\\\design\\\\icons')],
    hiddenimports=[],
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)
pyz = PYZ(a.pure, a.zipped_data, cipher=block_cipher)

exe = EXE(
    pyz,
    a.scripts,
    a.binaries,
    a.zipfiles,
    a.datas,
    [],
    exclude_binaries=False,
    name='App',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)
```

**Comment by rokm** [comment]
> So, the question is how can I create standalone executable without the additional step with the build tools installation?

You find out what DLL is missing, and ensure it is collected.

**Reply by hrimov** [reply]
> You find out what DLL is missing, and ensure it is collected.

So, for example I found, that it relates to the certain `.pyd`, then I need to include it via binaries-section?

**Reply by rokm** [reply]
I think the problem is not the .pyd itself, but one of its dependencies.

**Reply by hrimov** [reply]
I tried to debug it with the tool called Dependencies. But I cannot even understand, how should I include and what exactly I should include to the bundle. Here python310.dll is not defined, but with it everything is ok in the result bundle.

![open3d-dependencies](https://github.com/pyinstaller/pyinstaller/assets/45794357/73377c33-0c94-4617-8976-56d1ca440e0b)

**Reply by rokm** [reply]
Hmm, indeed. I've reproduced the problem with my Windows VMs, and I don't see any obvious dependency issues. Which means that the issue is a bit more subtle, and will require some more digging around. Will let you know what I find.

**Reply by rokm** [reply, marked as answer]
It turns out that in my case, it was not dependency issue; instead, I had AVX and AVX2 support/pass-through disabled on the "minimal" VM where I was testing (while it was enabled on the build VM).

So I guess to rule this out as a possible cause of the issue - what kind of machine are you trying to run the frozen program on? Does it support AVX/AVX2?

**Reply by hrimov** [reply]
@rokm Sorry for the late reply, is it a needed option - `Enable Nested VT-x /AMD-V`? Also, I have an intermediate result - I was succeed with the one of the build with copied all the DLL's from the source system and included them via binaries. So, I will try to reduce the needed binaries and let you know.
![photo_2024-02-06 13 50 34](https://github.com/pyinstaller/pyinstaller/assets/45794357/e4930f97-b109-4458-b33b-83cd0995b1f0)

**Reply by rokm** [reply]
I don't think so - that option is currently greyed-out on my Intel-based systems as well (that on linux host, though). 

You can use CPU-Z or a similar CPUID program to check what instruction sets are available within the VM.

**Reply by rokm** [reply]
And you can also search for "AVX" in the VM's log file, for entries like:
```
00:00:01.266459 Features
00:00:01.266459   Mnemonic - Description                                  = guest (host)
...
00:00:01.266506   AVX - AVX support                                       = 1 (1)
...
00:00:01.266513   AVX2 - Advanced Vector Extensions 2                     = 1 (1)
...
```

**Reply by hrimov** [reply]
@rokm Sorry for the disappearance, I lost access to the virtual machines for a while, the reason was with it, thank you for your time.
